### PR TITLE
Network refactor step1

### PR DESCRIPTION
**Pull request by trevelyan**
This push simplifies the structure of the Network in Saito-Lite-Rust.

**The Peer Class**
Peers are much simpler. They no longer have long and complicated internal functions that do random things like generate cryptographic challenges, initialize handshakes. Instead, communications are reduced to two main functions. The first of these is used to send messages and commands to other nodes (i.e. SENDMESG, REQBLOCK), while the second is used to send a reply when you are processing an inbound request or instruction:

 - peer.sendRequest(message, data = "")
 - peer.sendResponse()
 
So we are now handling ALL outbound messages using peer.sendRequest(). The function will check to see if it is a main / binary API function request and send those first, before falling back to a "default" SENDMESG transmission if there are no matches. Updating the network API requires adding new commands explicitly to this function.
 
 **The Network Class**
The network class is also simplified. Websockets are all initialized in network.initializeWebSocket() and the handshake and the request for syncing the blockchain is triggered here as well. Most of the random functions that were used to generate handshake challenges and support  the blockchain are also gone. In their place is a single function that catches requests that have been sent from other peers:

 - network.receivePeerRequest()
 
 I've cleaned up the way this works so that we are using a switch/case statement instead of nested if/else logic. In a lot of cases the existing code would almost randomly trigger functions in network / peer / networkapi or the files in the /networking folder. Now things are cleaner -- if what needs to be done can be handled in a reasonable amount of code it is tackled inline. There are several unused / legacy functions that are still there and which need to be removed. I would like feedback on structure before polishing these.
 
Sending messages to the network no longer needs nodes to interact with the peer class. Everything is handled through the network. These are the critical functions that nodes use to communicate with their peers:
 
 - network.fetchBlock()
 - network.propagateBlock()
 - network.propagateTransaction()
 - network.requestBlockchain()
 
On a more fundamental level, these functions can also be called manually to broadcast messages:
 
 - network.sendRequest()
 - network.sendPeerRequest()
 - network.sendRequestWithCallback()
 
The most important of these is the new function network.sendPeerRequest(). This function is used in most of the functions above that actually do something. All three of these functions find the relevant peers (if provided) and then pass control into the peer which broadcasts it using peer.sendRequest() [ see above ].

The result is a simplified set of functions with easier names. We are either SENDING requests or RECEIVING requests or SENDING responses. The network API file still exists, but it has shrunk down to only four functions. It's primary value is creating the promises that are used to handle chained execution.

**The Networking Folder**
Most of these files can be removed. Serialization and deserialization of simple messages can happen directly in whatever function sends the message and network.receivePeerRequest(). While this moves serialization and deserialization out of an ancillary file, I believe the simpler code structure will make it easier to develop. Fewer files. Fewer concepts. And tighter language.

**The Handshake**
There is a new file called handshake.js that handles the handshake. It will need to be upgraded as it is not validating signatures. We can add that in shortly. I have removed unnecessary data fields like the IP address as we can add them in later once everything is working.

Thoughts and feed

### CODE REVIEW

File: bundler/default/lib/saito/network.js [hunk 1]
```diff
@@ -159,18 +169,146 @@ class Network {
...
+
+
+    /**
+     * @param {string} block_hash
+     * @param {string} preferred peer (if exists); // TODO - remove duplicate function and update blockchain.js
+     */
+    async requestMissingBlock(block_hash, peer = null) { return await fetchBlock(block_hash, peer); }
+    async fetchBlock(block_hash, peer=null) {
+
+	if (peer == null) {
```
Comment (thaodt) [inline bundler/default/lib/saito/network.js:197]: is it required strictly to compare with `null` value? If yes, I want to suggest use strict equality operator `===`

File: lib/saito/peer.js [hunk 2]
```diff
@@ -135,409 +123,66 @@ class Peer {
...
-        this.socket.peer = this;
-
-        //
-        // TODO - test -- where is publickey set
-        //
-        //console.log("start emit handshake complete");
-        //this.app.connection.emit("handshake_complete", this);
-        //console.log("done emitting handshake complete");
-
+	if (this.app.BROWSER == 1) {
```
Comment (thaodt) [inline lib/saito/peer.js:126]: i think this `if else` can be simplified as below:
`this.socket = await this.app.network.initializeWebSocket(this, false, this.app.BROWSER == 1);`

File: bundler/default/lib/saito/network.js (hunk 1 above)
Comment (trevelyan) [reply bundler/default/lib/saito/network.js:197]: good idea. will push fix.

we're using operator overloading to simplify API:

  network.fetchBlock(block_hash);
  network.fetchBlock(block_hash, peer);
  
will cut "requestMissingBlock" once Sanka has pushed through the typescript update and I can wander into the blockchain file and fix that.

File: lib/saito/peer.js (hunk 2 above)
Comment (trevelyan) [reply lib/saito/peer.js:126]: beautiful

File: lib/saito/network.js [hunk 3]
```diff
@@ -437,6 +774,14 @@ class Network {
         }
     }
 
+    sendPeerRequestWithCallback(message, data = "", peer) {
+        for (let x = this.peers.length - 1; x >= 0; x--) {
+            if (this.peers[x] === peer) {
+                this.peers[x].sendRequestWithCallback(message, data, callback);
```
Comment (thaodt) [inline lib/saito/network.js:780]: it looks weird because we dont have `callback` here.

File: lib/saito/network.js [hunk 4]
```diff
@@ -437,6 +774,14 @@ class Network {
...
+            }
+        }
+    }
+
     sendRequest(message, data = "") {
```
Comment (thaodt) [inline lib/saito/network.js:785]: for 4 functions here `sendPeerRequest`, `sendPeerRequestWithCallback`, `sendRequest` & `sendRequestWithCallback`, I think we can simplify just 2 functions as below:
```
sendRequest(message, data = "",callback) {
  for (let x = this.peers.length - 1; x >= 0; x--) {
      callback ? this.peers[x].sendRequest(message, data) : this.peers[x].sendRequest(message, data, callback);
  }
}

sendPeerRequest(message, data = "", peer, isCallback) {
  for (let x = this.peers.length - 1; x >= 0; x--) {
    if (this.peers[x] === peer) {
        isCallback ? this.peers[x].sendRequest(message, data) : this.peers[x].sendRequest(message, data,callback);
    }
  }
}
```

**Comment by trevelyan** [comment]
Yes.

We can remove that function completely.

Thank you for flagging.

**Comment by trevelyan** [comment]
On second though, we need nothing more than:

 - sendRequest()
 - sendRequestWithCallback()

Had kept them untouched for backwards compatibility purposes, but we can
simply add peer as a final argument and update. Will do.

Thanks Thao.

File: bundler/default/lib/saito/network.js (hunk 1 above)
Comment (SankaD) [reply bundler/default/lib/saito/network.js]: if (!peer) is the way to go.

Comment (SankaD) [reply bundler/default/lib/saito/network.js]: if (!peer) is the way to go.

**Comment by SankaD** [comment]
@trevelyan I guess we can delete this PR now? since it's merged manually?

**Pull request closed without merging**